### PR TITLE
Interpreter: Mock exchange executors.

### PR DESCRIPTION
**Pull request by ywqzzy**
### What problem does this PR solve?

Issue Number: ref #4609 

Problem Summary:
I didn't support mock exchange executors function, in order to fully supprt mock dagrequest, need to implement all of them.

### What is changed and how it works?
Mock exchangeSender&Receivers.
Little refactor of Interpreter test utils in order to mock exchange executors easily.
### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Debug/astToExecutor.cpp [hunk 1]
```diff
@@ -1525,4 +1525,18 @@ ExecutorPtr compileJoin(size_t & executor_index, ExecutorPtr left, ExecutorPtr r
     return join;
 }
 
+
+ExecutorPtr compileExchangeSender(ExecutorPtr input, size_t & executor_index)
+{
+    ExecutorPtr exchange_sender = std::make_shared<mock::ExchangeSender>(executor_index, input->output_schema, tipb::PassThrough);
```
Comment (SeaRise) [inline dbms/src/Debug/astToExecutor.cpp:1531]: In many cases it is not tipb::PassThrough, but tipb::HashXXX.
need to be specified by the user

File: dbms/src/TestUtils/mockExecutor.cpp [hunk 2]
```diff
@@ -256,4 +290,8 @@ DAGRequestBuilder MockDAGRequestContext::scan(String db_name, String table_name)
     return DAGRequestBuilder(index).mockTable({db_name, table_name}, mock_tables[db_name + "." + table_name]);
 }
 
+DAGRequestBuilder MockDAGRequestContext::receive(String db_name, String table_name)
```
Comment (SeaRise) [inline dbms/src/TestUtils/mockExecutor.cpp:293]: It's a little weird to use `db_name` and `table_name`...

File: dbms/src/TestUtils/mockExecutor.h [hunk 3]
```diff
@@ -88,6 +94,7 @@ class DAGRequestBuilder
 
     ExecutorPtr root;
     DAGProperties properties;
+    std::unordered_map<String, std::vector<Int64>> receiver_source_task_ids_map;
```
Comment (SeaRise) [inline dbms/src/TestUtils/mockExecutor.h:97]: What is its use?

File: dbms/src/Debug/astToExecutor.cpp (hunk 1 above)
Comment (ywqzzy) [reply dbms/src/Debug/astToExecutor.cpp:1531]: I will fix it

File: dbms/src/TestUtils/mockExecutor.cpp (hunk 2 above)
Comment (ywqzzy) [reply dbms/src/TestUtils/mockExecutor.cpp:293]: I will save the relationship between sender and receiver.

**Review by yibin87 — APPROVED** [review]
LGTM

File: dbms/src/TestUtils/mockExecutor.h [hunk 4]
```diff
@@ -21,6 +21,8 @@
 #include <initializer_list>
 #include <unordered_map>
 
+#include "tipb/executor.pb.h"
```
Comment (ywqzzy) [inline dbms/src/TestUtils/mockExecutor.h:24]: I will fix it later.

File: dbms/src/TestUtils/InterpreterTestUtils.cpp [hunk 5]
```diff
@@ -47,9 +47,11 @@ String toTreeString(const tipb::Executor & root_executor, size_t level)
     FmtBuffer buffer;
 
     auto append_str = [&buffer, &level](const tipb::Executor & executor) {
-        assert(executor.has_executor_id());
```
[the remark this answers is not in the document]
Comment (ywqzzy) [reply dbms/src/TestUtils/InterpreterTestUtils.cpp:50]: Just a work around to mock senders which do not have children, I will fix it.

File: dbms/src/TestUtils/mockExecutor.cpp [hunk 6]
```diff
@@ -251,9 +284,30 @@ void MockDAGRequestContext::addMockTable(const MockTableName & name, const MockC
...
 }
 
+void MockDAGRequestContext::addExchangeRelationSchema(String name, const MockColumnInfos & columns)
+{
+    exchange_schemas[name] = columns;
+}
+
+void MockDAGRequestContext::addExchangeRelationSchema(String name, const MockColumnInfoList & columns)
+{
+    std::vector<MockColumnInfo> v_column_info;
```
Comment (Lloyd-Pottiger) [inline dbms/src/TestUtils/mockExecutor.cpp:294]: consider pre-allocating the container capacity before the loop?

File: dbms/src/TestUtils/mockExecutor.cpp [hunk 7]
```diff
@@ -67,13 +69,13 @@ void DAGRequestBuilder::initDAGRequest(tipb::DAGRequest & dag_request)
...
 
 // traval the AST tree to build tipb::Executor recursively.
-std::shared_ptr<tipb::DAGRequest> DAGRequestBuilder::build(Context & context)
+std::shared_ptr<tipb::DAGRequest> DAGRequestBuilder::build(MockDAGRequestContext & mock_context)
 {
-    MPPInfo mpp_info(properties.start_ts, -1, -1, {}, {});
+    MPPInfo mpp_info(properties.start_ts, -1, -1, {}, mock_context.receiver_source_task_ids_map);
     std::shared_ptr<tipb::DAGRequest> dag_request_ptr = std::make_shared<tipb::DAGRequest>();
     tipb::DAGRequest & dag_request = *dag_request_ptr;
+    root->toTiPBExecutor(dag_request.mutable_root_executor(), properties.collator, mpp_info, mock_context.context);
```
Comment (SeaRise) [inline dbms/src/TestUtils/mockExecutor.cpp:77]: why move this line to here?

**Review by SeaRise — APPROVED** [review]

File: dbms/src/TestUtils/mockExecutor.cpp (hunk 6 above)
Comment (ywqzzy) [reply dbms/src/TestUtils/mockExecutor.cpp:294]: Good idea！

**Review by Lloyd-Pottiger — APPROVED** [review]
LGTM, cool!

**Pull request merged**